core/config/config_helper.cpp

```diff
@@ -55,9 +57,101 @@ parse_or_get_factory<const stop::CriterionFactory>(const pnode& config,
     }
+
+    GKO_INVALID_STATE(
+        "Criteria must either be defined as a string or an array.");
+}
+
+
+std::vector<deferred_factory_parameter<const stop::CriterionFactory>>
+parse_minimal_criteria(const pnode& config, const registry& context,
+                       const type_descriptor& td)
+{
+    auto map_time = [](const pnode& config, const registry& context,
+                       const type_descriptor& td) {
+        pnode time_config{{{"time_limit", config.get("time")}}};
+        return configure_time(time_config, context, td);
+    };
+    auto map_iteration = [](const pnode& config, const registry& context,
+                            const type_descriptor& td) {
+        pnode iter_config{{{"max_iters", config.get("iteration")}}};
+        return configure_iter(iter_config, context, td);
+    };
+    auto create_residual_mapping = [](const std::string& key,
+                                      const std::string& baseline,
+                                      auto configure_fn) {
+        return std::make_pair(
+            key, [=](const pnode& config, const registry& context,
+                     const type_descriptor& td) {
+                pnode res_config{{{"baseline", pnode{baseline}},
+                                  {"reduction_factor", config.get(key)}}};
+                return configure_fn(res_config, context, td);
+            });
+    };
+    std::map<
+        std::string,
+        std::function<deferred_factory_parameter<gko::stop::CriterionFactory>(
+            const pnode&, const registry&, type_descriptor)>>
+        criterion_map{
+            {{"time", map_time},
+             {"iteration", map_iteration},
+             create_residual_mapping("relative_residual_norm", "rhs_norm",
+                                     configure_residual),
+             create_residual_mapping("initial_residual_norm", "initial_resnorm",
+                                     configure_residual),
+             create_residual_mapping("absolute_residual_norm", "absolute",
+                                     configure_residual),
+             create_residual_mapping("relative_implicit_residual_norm",
+                                     "rhs_norm", configure_implicit_residual),
+             create_residual_mapping("initial_implicit_residual_norm",
+                                     "initial_resnorm",
+                                     configure_implicit_residual),
+             create_residual_mapping("absolute_implicit_residual_norm",
+                                     "absolute", configure_implicit_residual)}};
+
+    auto value_type_it = config.get_map().find("value_type");
+    type_descriptor updated_td =
+        value_type_it == config.get_map().end()
+            ? td
+            : type_descriptor(value_type_it->second.get_string(),
+                              td.get_local_index_typestr(),
+                              td.get_global_index_typestr());
```

Comment:
you can also use `updated_td = config::update_type(td);`

Reply:
Thanks for the hint!